src/porepy/numerics/ad/equation_system.py

```diff
@@ -1035,31 +997,40 @@ def dofs_of(self, variables: VariableList) -> np.ndarray:
             variables: VariableType input for which the indices are requested.
 
         Returns:
-            an order-preserving array of indices of DOFs belonging to the VariableType input.
+            An order-preserving array of indices of DOFs corresponding to ``variables``.
 
         Raises:
             ValueError: if unknown VariableType arguments are passed.
 
         """
         variables = self._parse_variable_type(variables)
+        var_ids = [var.id for var in variables]
         global_variable_dofs = np.hstack((0, np.cumsum(self._variable_num_dofs)))
 
         # Storage of indices per requested variable.
+        # NOTE Loop over current order in variable numbers to preserve order of dofs
         indices = list()
-        for variable in variables:
```

Comment:
dof_of returned the indices depending on the order of the VariableList argument.
Fixed s.t. dofs are always ascending (reflecting order imposed by equation system).

Reply:
The output was an array with indices, simply concatenating indices of variables found in the input argument.
So the output was sensitive to the order of the input arguments.
Now, the method loops over the current dof order and computes indices for each variable in the input argument.
The output is always ordered ascending and reflects the DOF order.

Reply:
I have been so far at most going via the private methods of the `DiagnosticsMixin,` like `_variable_data` - very useful and possible worth extracting somewhere or making it somewhat publicly available TBC. See below for an example code snippet:
![image](https://github.com/pmgbergen/porepy/assets/73653591/9c67e128-0f53-4b77-9ab3-47c4e32ff8f6)

Reply:
I do not rely on the order returned from `dofs_of`. Instead, I store and manipulate the indices of variables outside the model. The `dofs_of` is used only to retreat the indices associated with an individual MD-variable.

Reply:
I suggest the following resolutions:
* The variable id is considered a private field (conversion to _id can be postponed, since work on other PRs should have higher priority). The user is expected to identify variables in terms of its name and domain.
* Internally to the EquationSystem, we are free to organize variable lookup and ordering of dofs as we best see fit. I believe the changes introduced in this PR is this regard are sound, so I'm mostly fine with the changes herein.
* The method dofs_of is the way the user can fetch the global DOFs for a list of variables. The user does not know the order the EqSys has chosen for the variables, thus this method should give the dofs in the order of the input list of variables. That is, *the changes to this specific function should be reverted*.

Does this make sense, @vlipovac?

I will document more clearly the intention for the Variable.id and push a small commit to this branch.

Reply:
I turned the ID into a property, with a private `_id`.
Users have access to the public `id`, without being able to modify it.